languages/ja/nodemaker.lua

```diff
@@ -0,0 +1,340 @@
+local badAfterClasses = {}
+for _, v in ipairs({ 2, 3, 4, 5, 6, 7, 9, 10, 11, 20, 29 }) do
+   badAfterClasses[v] = true
+end
+
+local function breakAllowed (before, after)
+   local bc = jisClass(before)
+   local ac = jisClass(after)
+   if badBeforeClasses[bc] then
+      return false
+   end
+   if badAfterClasses[ac] then
+      return false
+   end
+   if bc == 8 and ac == 8 then
+      if before == after then
+         return false
+      else
+         return true
+      end
+   end
+   -- The following rule is not in jlreq but it stops the algorithm from breaking
+   -- up Latin words!
+   if bc == 27 and ac == 27 then
+      return false
+   end
+   return true
+end
+
+local function intercharacterspace (before, after)
+   local bc = jisClass(before)
+   local ac = jisClass(after)
+   -- This rule is not in jlreq but it stops situations like 1：2 getting munched
+   if (bc == 5 and ac == 27) or (bc == 27 and ac == 5) then
+      return 0
+   end
+
+   -- jlreq expects yakumono to contain no following space, so recommends half
+   -- a space afterwards. However, in real world fonts, yakumono occupies a full
+   -- zenkaku width with trailing space built into the glyph. So we do not add
+   -- space there. Instead, where no space is stipulated, we add negative space
+   -- to counteract the trailing space in the glyph.
+   if ac == 5 then
+      return "-0.25zw"
+   end
+   if bc == 5 then
+      return "-0.25zw"
+   end
+   if bc == 7 and ac == 1 then
+      return "-0.5zw"
+   end
+   if bc == 2 and ac == 1 then
+      return "-0.5zw"
+   end
+   if bc == -1 and ac == 1 then
+      return "-0.5zw"
+   end
+   if bc == 6 or bc == 7 then
+      if ac == 2 or ac == 6 or ac == 7 then
+         return "-0.25zw"
+      end
+      return 0
+   end
+   if (bc == 9 or bc == 10 or bc == 11 or bc == 15 or bc == 16 or bc == 19) and (ac == 21 or ac == 24 or ac == 25) then
+      return "0.25zw"
+   end
+
+   return 0
+end
+
+local function stretchability (before, after)
+   local bc = jisClass(before)
+   local ac = jisClass(after)
+   -- somewhat simplified from table 6 of jlreq
+   if ac == 1 then
+      return "0.25zw"
+   end
+   if ac < 8 then
+      return 0
+   end
+   if bc == 4 and (ac == 21 or ac == 24 or ac == 25 or ac == 27) then
+      return "0.5zw"
+   end
+   if bc < 8 then
+      return 0
+   end
+   return "0.25zw" -- somewhat simplified
+end
+
+local function shrinkability (before, after)
+   local bc = jisClass(before)
+   local ac = jisClass(after)
+   -- This rule is not in jlreq but it stops situations like 1：2 getting munched
+   if (bc == 5 and ac == 27) or (bc == 27 and ac == 5) then
+      return 0
+   end
+   -- somewhat simplified from table 5 of jlreq
+   if ac == 1 then
+      return "0.5zw"
+   end
+   if ac == 5 then
+      return "0.25zw"
+   end
+   if bc == 5 then
+      return "0.25zw"
+   end
+   if bc == 7 then
+      return "0.5zw"
+   end
+   return 0
+end
+
+-- local okbreak = SILE.types.node.penalty(0)
```

Comment:
Present in the previous code base, but what is it?

Reply:
I don't know, a smudgy leftover from early development. It look like it was actually used in 2aaf16a9f then ditched in 1e2e4a40d and never actually used again.